### PR TITLE
Builds perfectly on windows (Visual studio code) and added Makefile(make).

### PR DESCRIPTION
**Pull request by DivyPatel9881**

### CODE REVIEW

**Comment by DivyPatel9881** [comment]
@hsluoyz you can review the code and pull the commits.

File: src/rbac/DefaultRoleManager.h [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                if ((*it).hasRole(name, hierarchyLevel - 1)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        bool hasDirectRole(string name) {
+            vector <Role> :: iterator it;
```
Comment (xcaptain) [inline src/rbac/DefaultRoleManager.h]: how about using auto keyword so we don't need to declare it outside for loop

File: src/rbac/DefaultRoleManager.h [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                names.push_back((*it).name);
+            }
+            return names;
+        }
+};
+
+class DefaultRoleManager : public RoleManager {
+    private:
+        map <string, Role> allRoles;
```
Comment (xcaptain) [inline src/rbac/DefaultRoleManager.h]: use `map` or `unordered_map`?

Comment (DivyPatel9881) [reply src/rbac/DefaultRoleManager.h]: Yeah will use unordered_map here as the order of the keys won't matter here

File: src/effect/Effect.h [hunk 3]
```diff
@@ -0,0 +1,5 @@
+enum Effect{
```
[the remark this answers is not in the document]
Comment (DivyPatel9881) [reply src/effect/Effect.h]: Yeah, will make changes.

File: src/rbac/DefaultRoleManager.h (hunk 1 above)
Comment (DivyPatel9881) [reply src/rbac/DefaultRoleManager.h]: Actually we need verbosity in code, which auto will kill, but will define the iterator inside the loop.

File: src/rbac/DefaultRoleManager.h [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+#include "exception/IllegalArgumentException.h"
+#include "log/LogUtil.h"
+
+/**
+ * Role represents the data structure for a role in RBAC.
+ */
+class Role {
+    
+    private:
+        vector <Role> roles;
```
[the remark this answers is not in the document]
Comment (DivyPatel9881) [reply src/rbac/DefaultRoleManager.h]: Its illegal to put this references in types, but I think code will work just fine.

File: src/exception/IllegalArgumentException.h [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#include <string>
+
+// Exception class for unsupported operations.
+class IllegalArgumentException{
+    std::string error_message;
+    public:
+        IllegalArgumentException(std::string error_message){
```
[the remark this answers is not in the document]
Comment (DivyPatel9881) [reply src/exception/IllegalArgumentException.h]: Actually, I have used initializer list whereever necessary, but have avoided where I don't have to. for example, I have used the same when I have to initialise the super attributes, but not for class member attributes. Using initializer list decreses readability of the code.

File: src/rbac/DefaultRoleManager.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (DivyPatel9881) [reply src/rbac/DefaultRoleManager.h]: Thats illegal, writing reference types. The existing code will work just fine.

**Comment by DivyPatel9881** [comment]
@hsluoyz can review and merge it?

**Comment by hsluoyz** [comment]
Please rework/rebase/remove the wrong commits:

image

**Comment by DivyPatel9881** [comment]
Have squashed the commits to the important ones, now you can merge it.

**Comment by hsluoyz** [comment]
Make your code compile on Visual Studio.

**Comment by DivyPatel9881** [comment]
I have made a Makefile and the codebase builds perfectly on my system(Windows Visual Studio Code.)

**Comment by DivyPatel9881** [comment]
@hsluoyz can you please confirm it on your working environment and let me know if any changes to be made on my side or feel free to change.

**Comment by hsluoyz** [comment]
Base your code on our latest master branch, follow our folder structure, filename naming, etc.

https://github.com/casbin/casbin-cpp/issues/7

**Comment by DivyPatel9881** [comment]
Yeah, already done and also following Google CPP code style guide.

**Pull request closed without merging**